### DISCUSSION TITLE
Using of type selector in themes

### DISCUSSION THREAD

**Discussion by RomanSoloweow** [AvaloniaUI/Avalonia#6353] [General]
Most of the styles are written with a direct selector (without using "is"). This becomes a problem when inheriting from base controls.

Why is this not fixed for Avalonia.Themes.Default and Avalonia.Themes.Fluent?

**Comment by grokys** [comment]
This is because there's no way to opt-out of an `:is` selector. If we used `:is` in the themes then that theme would apply to all controls derived from that control _whether you want it or not_.

For this reason if you want to use an inherited control's theme you must opt-in by implementing the `IStyleable.StyleKey` property, for example:

```csharp
class MyButton : Button, IStyleable
{
    Type IStyleable.StyleKey => typeof(Button);
}
```

WPF/UWP requries a similar opt-in but the mechanism is different - there it's done by [setting the`DefaultStyleKey` property](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.defaultstylekey?redirectedfrom=MSDN&view=net-5.0#System_Windows_FrameworkElement_DefaultStyleKey)

**Reply by maxkatz6** [reply]
> For this reason if you want to use an inherited control's theme you must opt-in by implementing the IStyleable.StyleKey property, for example:

There is a problem with this approach, that you can't in the same time extend default styles (from Button type in your example), but also add something specifically to MyButton.
Although it's possible to add additional class name to solve it.
I also not sure, but probably our :is() ignores actual type, when StyleKey is defined. Should be double checked.

**Reply by RomanSoloweow** [reply]
@maxkatz6  is right by setting StyleKey => typeof (Button), I won't be able to write styles for MyButton and this is a real problem.

**Reply by maxkatz6** [reply]
@RomanSoloweow I wrote a workaround above

**Reply by RomanSoloweow** [reply]
> Although it's possible to add additional class name to solve it.

Yes, this is bypassed with a class name that you can even add directly in the MyButton constructor. But this is more a trick than obvious and expected behavior.

**Reply by maxkatz6** [reply]
Does :is(MyButton) works for you?

**Reply by RomanSoloweow** [reply]
I've tried this with ItemsControl. None of the selectors work with StyleKey => typeof (ItemsControl):

Selector = ": is (controls | MyItemsControl)"
Selector = "controls | MyItemsControl"

**Reply by graealex** [reply]
> Although it's possible to add additional class name to solve it.

I also experienced this issue, and a big problem with differentiating by class is that the Style using Selector="Button" won't have access to custom properties implemented in the type MyButton - if any. For that, you need to reference the style via Selector="MyButton" or Selector=":is(MyButton)" - but then it isn't applied anymore. There should either be another selector-kind, or :is() should look at the actual type, and not the StyleKey.
